src/lasse_handler.erl

```diff
@@ -20,186 +21,190 @@
-    Module = State#state.module,
-    ModuleState = State#state.state,
-    Module:terminate(Reason, Req, ModuleState),
-    ok.
+terminate(Reason, Req, State) ->
+  Module = State#state.module,
+  ModuleState = State#state.state,
+  Module:terminate(Reason, Req, ModuleState),
+  ok.
+
+-spec upgrade(Req, Env, module(), any(), timeout(), run | hibernate)
+  -> {ok, Req, Env} | {suspend, module(), atom(), [any()]}
+  when Req::cowboy_req:req(), Env::cowboy_middleware:env().
+upgrade(Req, Env, Handler, HandlerState, Timeout, run) ->
+  cowboy_loop:upgrade(Req, Env, Handler, HandlerState, Timeout, run);
+upgrade(Req, Env, Handler, HandlerState, Timeout, hibernate) ->
+  cowboy_loop:upgrade(Req, Env, Handler, HandlerState, Timeout, hibernate).
 
 %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
 %%% API
 %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
 -spec notify(atom() | pid(), term()) -> ok.
 notify(Pid, Msg) ->
-    Pid ! {message, Msg},
-    ok.
+  Pid ! {message, Msg},
+  ok.
 
 %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
 %% Helper functions
 %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
 handle_init({ok, Req, State}, Module) ->
-    handle_init({ok, Req, [], State}, Module);
+  handle_init({ok, Req, [], State}, Module);
 handle_init({ok, Req, InitialEvents, State}, Module) ->
-    case cowboy_req:method(Req) of
-        {<<"GET">>, Req1} ->
-            % "no-cache recommended to prevent caching of event data.
-            Headers = [{<<"content-type">>, <<"text/event-stream">>},
-                       {<<"cache-control">>, <<"no-cache">>}],
-            {ok, Req2} = cowboy_req:chunked_reply(200, Headers, Req1),
-
-            lists:foreach(
-              fun(Event) -> ok = send_event(Event, Req2) end,
-              InitialEvents
-             ),
-
-            {loop, Req2, #state{module = Module, state = State}};
-        {_OtherMethod, _} ->
-            Headers = [{<<"content-type">>, <<"text/html">>}],
-            StatusCode = 405, % Method not Allowed
-            {ok, Req1} = cowboy_req:reply(StatusCode, Headers, Req),
-            {shutdown, Req1, #state{module = Module}}
-    end;
+  case cowboy_req:method(Req) of
+    <<"GET">> ->
+      % "no-cache recommended to prevent caching of event data.
+      Headers = [{<<"content-type">>, <<"text/event-stream">>},
+                 {<<"cache-control">>, <<"no-cache">>}],
+      Req1 = cowboy_req:chunked_reply(200, Headers, Req),
+      lists:foreach(
+        fun(Event) -> ok = send_event(Event, Req1) end,
+        InitialEvents
+      ),
+      {cowboy_loop, Req1, #state{module = Module, state = State}};
+
+   _OtherMethod ->
+      Headers = [{<<"content-type">>, <<"text/html">>}],
+      StatusCode = 405, % Method not Allowed
+      Req1 = cowboy_req:reply(StatusCode, Headers, Req),
+      {cowboy_loop, Req1, #state{module = Module}}
```

Comment:
You replaced `shutdown` by `cowboy_loop`. Is that expected, how are you telling cowboy to just return to the client without actually looping? Is that accomplished using `cowboy_req:reply/3` in the line above?